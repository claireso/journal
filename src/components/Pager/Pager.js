import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import * as S from './Pager.styles'

const Pager = ({ navigate, first, prev, next, last }) => {
  const handleClick = useCallback((page) => () => navigate(page), [navigate])

  return (
    <S.Wrapper>
      {first && (
        <li>
          <S.Button data-testid="first-page" title="First page" onClick={handleClick(first)}>
            ««
          </S.Button>
        </li>
      )}

      {prev && (
        <li>
          <S.Button data-testid="previous-page" title="Previous page" onClick={handleClick(prev)}>
            «
          </S.Button>
        </li>
      )}

      {next && (
        <li>
          <S.Button data-testid="next-page" title="Next page" onClick={handleClick(next)}>
            »
          </S.Button>
        </li>
      )}

      {last && (
        <li>
          <S.Button data-testid="last-page" title="Last page" onClick={handleClick(last)}>
            »
          </S.Button>
        </li>
      )}
    </S.Wrapper>
  )
}

Pager.propTypes = {
  navigate: PropTypes.func.isRequired,
  first: PropTypes.number,
  prev: PropTypes.number,
  next: PropTypes.number,
  last: PropTypes.number
}

export default Pager
