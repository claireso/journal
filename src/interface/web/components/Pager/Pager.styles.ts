import { styled } from '@web/oldtheme'

import { ButtonSecondary } from '../Buttons'

export const Wrapper = styled('ul', {
  display: 'inline-flex',
  justifyContent: 'center',
  listStyleType: 'none',
  width: '100%',
  p: 0,
  '> li': {
    mx: '0.4rem'
  }
})

export const Button = styled(ButtonSecondary, {
  alignItems: 'center',
  fontSize: '$4',
  height: '4.4rem',
  justifyContent: 'center',
  p: 0,
  width: '4.4rem'
})
