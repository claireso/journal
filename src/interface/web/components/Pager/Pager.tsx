import React, { useCallback } from 'react'

import { ButtonDark } from '../Buttons'

import * as cls from './styles.css'

interface PagerProps {
  navigate: (page: string) => void
  first?: number
  prev?: number
  next?: number
  last?: number
}

const Pager = ({ navigate, first, prev, next, last }: PagerProps) => {
  const handleClick = useCallback((page: number) => () => navigate(`${page}`), [navigate])

  return (
    <ul className={cls.wrapper}>
      {first && (
        <li>
          <ButtonDark title="First page" onClick={handleClick(first)}>
            ««
          </ButtonDark>
        </li>
      )}

      {prev && (
        <li>
          <ButtonDark title="Previous page" onClick={handleClick(prev)}>
            «
          </ButtonDark>
        </li>
      )}

      {next && (
        <li>
          <ButtonDark title="Next page" onClick={handleClick(next)}>
            »
          </ButtonDark>
        </li>
      )}

      {last && (
        <li>
          <ButtonDark title="Last page" onClick={handleClick(last)}>
            »»
          </ButtonDark>
        </li>
      )}
    </ul>
  )
}

export default Pager
