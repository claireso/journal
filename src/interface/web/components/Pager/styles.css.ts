import { globalStyle } from '@vanilla-extract/css'
import { cmpStyles, gap, square } from '@web/theme'

export const wrapper = cmpStyles({
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'center',
  ...gap('size-2')
})

globalStyle(`${wrapper} button`, {
  placeContent: 'center',
  ...square('size-11')
})
