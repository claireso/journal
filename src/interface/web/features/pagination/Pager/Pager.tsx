'use client'

import { useCallback } from 'react'
import { useRouter } from 'next/navigation'
import type { Pager } from '@domain/entities'
import _Pager from '@web/components/Pager'

interface PagerProps extends Pick<React.ComponentProps<typeof _Pager>, 'layout'> {
  pager: Pager
}

const Pager = ({ pager, layout }: PagerProps) => {
  const router = useRouter()

  const navigate = useCallback(
    (page: string) => {
      router.push(`?page=${page}`)
    },
    [router]
  )

  console.log({ pager })

  return <_Pager navigate={navigate} layout={layout} {...pager} />
}

export default Pager
