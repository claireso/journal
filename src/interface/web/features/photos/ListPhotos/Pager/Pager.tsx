'use client'

import { useCallback } from 'react'
import { useRouter } from 'next/navigation'
import { styled } from '@web/oldtheme'
import _Pager from '@web/components/Pager'

const PagerWrapper = styled('div', {
  gridColumn: '1 / -1',
  m: '4.5rem 0',
  '@lg': {
    m: '8.5rem 0'
  }
})

interface PagerProps {
  pager: {
    count?: number
    first?: number
    prev?: number
    next?: number
    last?: number
  }
}

const Pager = ({ pager }: PagerProps) => {
  const router = useRouter()

  const navigate = useCallback(
    (page: string) => {
      router.push(`/?page=${page}`)
    },
    [router]
  )

  return (
    <PagerWrapper>
      <_Pager navigate={navigate} {...pager}></_Pager>
    </PagerWrapper>
  )
}

export default Pager
