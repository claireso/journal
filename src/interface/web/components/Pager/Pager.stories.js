import React from 'react'
import { action } from '@storybook/addon-actions'

import Pager from './index'

const params = {
  title: 'Components/Pager'
}

export default params

export const Basic = () => {
  const data = {
    first: 1,
    last: 8,
    next: 3,
    prev: 1
  }

  return <Pager {...data} navigate={action('on page change')} />
}
