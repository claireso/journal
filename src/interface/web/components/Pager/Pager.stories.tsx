import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Pager from './Pager'

const meta: Meta<typeof Pager> = {
  title: 'Components/Pager',
  component: Pager
}

export default meta
type Story = StoryObj<typeof Pager>

export const Primary: Story = {
  args: {
    first: 1,
    last: 8,
    next: 3,
    prev: 1,
    navigate: action('on page change')
  }
}
