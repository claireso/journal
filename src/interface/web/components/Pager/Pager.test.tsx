// @ts-nocheck
import { render, screen, fireEvent } from '@testing-library/react'

import Pager from './index'

const renderPager = (props = {}) => render(<Pager navigate={jest.fn()} {...props} />)

describe('<Pager />', () => {
  it('should not render items', () => {
    const { asFragment } = renderPager({ navigate: () => {} })

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render first page', () => {
    const props = {
      navigate: jest.fn(),
      next: global.__PHOTOS__.pager.next,
      last: global.__PHOTOS__.pager.last
    }

    const { asFragment } = renderPager(props)

    expect(asFragment()).toMatchSnapshot()

    fireEvent.click(screen.getByTitle('Next page'))
    expect(props.navigate).toHaveBeenCalledWith('2')

    fireEvent.click(screen.getByTitle('Last page'))
    expect(props.navigate).toHaveBeenCalledWith('19')
  })

  it('should render second page', () => {
    const props = {
      navigate: jest.fn(),
      first: 1,
      prev: 1,
      next: global.__PHOTOS__.pager.next,
      last: global.__PHOTOS__.pager.last
    }

    const { asFragment } = renderPager(props)

    expect(asFragment()).toMatchSnapshot()

    fireEvent.click(screen.getByTitle('First page'))
    expect(props.navigate).toHaveBeenCalledWith('1')

    fireEvent.click(screen.getByTitle('Previous page'))
    expect(props.navigate).toHaveBeenCalledWith('1')

    fireEvent.click(screen.getByTitle('Next page'))
    expect(props.navigate).toHaveBeenCalledWith('2')

    fireEvent.click(screen.getByTitle('Last page'))
    expect(props.navigate).toHaveBeenCalledWith('19')
  })

  it('should render last page', () => {
    const props = {
      first: 1,
      prev: 1,
      navigate: jest.fn()
    }

    const { asFragment } = renderPager(props)

    expect(asFragment()).toMatchSnapshot()

    fireEvent.click(screen.getByTitle('First page'))
    expect(props.navigate).toHaveBeenCalledWith('1')

    fireEvent.click(screen.getByTitle('Previous page'))
    expect(props.navigate).toHaveBeenCalledWith('1')
  })
})
