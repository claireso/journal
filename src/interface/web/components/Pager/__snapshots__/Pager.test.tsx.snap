// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pager /> should not render items 1`] = `
<DocumentFragment>
  <ul
    class="styles_layerStyle__ltatc70"
  />
</DocumentFragment>
`;

exports[`<Pager /> should render first page 1`] = `
<DocumentFragment>
  <ul
    class="styles_layerStyle__ltatc70"
  >
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="Next page"
      >
        »
      </button>
    </li>
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="Last page"
      >
        »»
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<Pager /> should render last page 1`] = `
<DocumentFragment>
  <ul
    class="styles_layerStyle__ltatc70"
  >
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="First page"
      >
        ««
      </button>
    </li>
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="Previous page"
      >
        «
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<Pager /> should render second page 1`] = `
<DocumentFragment>
  <ul
    class="styles_layerStyle__ltatc70"
  >
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="First page"
      >
        ««
      </button>
    </li>
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="Previous page"
      >
        «
      </button>
    </li>
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="Next page"
      >
        »
      </button>
    </li>
    <li>
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_md__1mdc28jl styles_layerStyle__1mdc28j6"
        title="Last page"
      >
        »»
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
